File: proctoring/app/proctoring/eye_tracking.py
"""
Eye tracking and gaze direction detection
"""
import mediapipe as mp
import numpy as np
from typing import Tuple, Optional

from app.core.logging import logger


class EyeTracker:
    """Eye tracking and gaze direction using MediaPipe Face Mesh"""

    # Face Mesh landmark indices
    LEFT_EYE = [362, 385, 387, 263, 373, 380]
    RIGHT_EYE = [33, 160, 158, 133, 153, 144]

    # Gaze direction landmarks
    LEFT_IRIS = [468]
    RIGHT_IRIS = [473]

    def __init__(self):
        """Initialize MediaPipe Face Mesh"""
        self.mp_face_mesh = mp.solutions.face_mesh
        self.face_mesh = self.mp_face_mesh.FaceMesh(
            static_image_mode=False,
            max_num_faces=1,
            refine_landmarks=True,
            min_detection_confidence=0.7,
            min_tracking_confidence=0.5,
        )

    def detect_eyes_looking_away(
        self,
        image: np.ndarray,
        threshold: float = 0.3,
    ) -> Tuple[bool, Optional[dict]]:
        """
        Detect if eyes are looking away from camera.
        
        Args:
            image: OpenCV image array (BGR)
            threshold: Threshold for gaze direction (0-1)
            
        Returns:
            Tuple of (is_looking_away, gaze_data)
            gaze_data contains:
                'left_iris_x': normalized iris x position,
                'right_iris_x': normalized iris x position,
                'horizontal_gaze': -1 (left), 0 (center), 1 (right)
        """
        rgb_image = image[:, :, ::-1]
        results = self.face_mesh.process(rgb_image)

        if not results.multi_face_landmarks or len(results.multi_face_landmarks) == 0:
            return True, None  # No face = looking away

        landmarks = results.multi_face_landmarks[0].landmark
        h, w, _ = image.shape

        # Get iris positions
        left_iris = landmarks[self.LEFT_IRIS[0]]
        right_iris = landmarks[self.RIGHT_IRIS[0]]

        # Normalize iris positions relative to eye regions
        left_eye_points = [landmarks[i] for i in self.LEFT_EYE]
        right_eye_points = [landmarks[i] for i in self.RIGHT_EYE]

        left_eye_center_x = np.mean([p.x for p in left_eye_points])
        right_eye_center_x = np.mean([p.x for p in right_eye_points])

        # Calculate horizontal gaze direction
        left_gaze = left_iris.x - left_eye_center_x
        right_gaze = right_iris.x - right_eye_center_x

        # Determine if looking away
        looking_away = abs(left_gaze) > threshold or abs(right_gaze) > threshold

        gaze_data = {
            "left_iris_x": left_iris.x,
            "right_iris_x": right_iris.x,
            "left_gaze_offset": left_gaze,
            "right_gaze_offset": right_gaze,
            "horizontal_gaze": (
                -1 if (left_gaze < -threshold or right_gaze < -threshold) else
                1 if (left_gaze > threshold or right_gaze > threshold) else
                0
            ),
        }

        return looking_away, gaze_data

    def close(self):
        """Close tracker resources"""
        if self.face_mesh:
            self.face_mesh.close()

File: proctoring/src/proctoring/eyeTracking.js
/**
 * Eye tracking and gaze direction detection using face landmarks
 * Analyzes eye position to detect if student is looking away
 */
const logger = require('../config/logger');

// Gaze offset threshold for looking away detection
const GAZE_THRESHOLD = 0.25;

/**
 * Calculate eye aspect ratio (EAR) for blink/eye closure detection
 * @param {Array} eye - 6 landmark points for one eye
 * @returns {number} - Eye aspect ratio
 */
function calculateEAR(eye) {
  if (!eye || eye.length < 6) return 0.3; // Default to open eye

  // Vertical eye distances
  const v1 = Math.sqrt(Math.pow(eye[1].x - eye[5].x, 2) + Math.pow(eye[1].y - eye[5].y, 2));
  const v2 = Math.sqrt(Math.pow(eye[2].x - eye[4].x, 2) + Math.pow(eye[2].y - eye[4].y, 2));

  // Horizontal eye distance
  const h = Math.sqrt(Math.pow(eye[0].x - eye[3].x, 2) + Math.pow(eye[0].y - eye[3].y, 2));

  // EAR = (v1 + v2) / (2 * h)
  const ear = (v1 + v2) / (2 * h);
  return ear;
}

/**
 * Detect if eyes are looking away from camera using face landmarks
 * @param {Buffer} imageBuffer - Image buffer (unused, for API compatibility)
 * @param {Array} landmarks - Face landmarks (68 points)
 * @param {number} threshold - Threshold for gaze direction (0-1)
 * @returns {Object} - {lookingAway: boolean, gazeData: Object|null}
 */
function detectEyesLookingAway(imageBuffer, landmarks = null, threshold = GAZE_THRESHOLD) {
  try {
    // If no landmarks provided, we can't detect gaze
    if (!landmarks || landmarks.length === 0) {
      logger.warning('No landmarks for eye tracking');
      return { lookingAway: false, gazeData: null };
    }

    // Get first face landmarks if array of arrays
    const faceLandmarks = Array.isArray(landmarks[0]) ? landmarks[0] : landmarks;

    if (faceLandmarks.length < 68) {
      logger.warning(`Insufficient landmarks for eye tracking: ${faceLandmarks.length}`);
      return { lookingAway: false, gazeData: null };
    }

    // Get left eye landmarks (36-41) and right eye landmarks (42-47)
    const leftEye = faceLandmarks.slice(36, 42);
    const rightEye = faceLandmarks.slice(42, 48);

    // Calculate eye centers
    const leftEyeCenter = {
      x: leftEye.reduce((sum, p) => sum + p.x, 0) / leftEye.length,
      y: leftEye.reduce((sum, p) => sum + p.y, 0) / leftEye.length,
    };
    const rightEyeCenter = {
      x: rightEye.reduce((sum, p) => sum + p.x, 0) / rightEye.length,
      y: rightEye.reduce((sum, p) => sum + p.y, 0) / rightEye.length,
    };

    // Get face center (nose bridge - landmark 27-30)
    const noseBridge = faceLandmarks[27];
    const noseTip = faceLandmarks[30];

    // Calculate horizontal gaze offset
    // Eyes should be roughly symmetric around nose bridge
    const avgEyeX = (leftEyeCenter.x + rightEyeCenter.x) / 2;
    const eyeDistance = rightEyeCenter.x - leftEyeCenter.x;

    // Horizontal offset of eye centers relative to nose
    const horizontalOffset = (avgEyeX - noseBridge.x) / eyeDistance;

    // Calculate vertical gaze (are eyes looking up/down based on relative positions)
    const avgEyeY = (leftEyeCenter.y + rightEyeCenter.y) / 2;
    const verticalOffset = (avgEyeY - noseBridge.y) / eyeDistance;

    // Calculate EAR for blink detection
    const leftEAR = calculateEAR(leftEye);
    const rightEAR = calculateEAR(rightEye);
    const avgEAR = (leftEAR + rightEAR) / 2;

    const gazeData = {
      left_eye_center: leftEyeCenter,
      right_eye_center: rightEyeCenter,
      horizontal_gaze: Math.round(horizontalOffset * 100) / 100,
      vertical_gaze: Math.round(verticalOffset * 100) / 100,
      eye_aspect_ratio: Math.round(avgEAR * 100) / 100,
      eyes_closed: avgEAR < 0.2,
    };

    // Determine if looking away
    const lookingAway = Math.abs(horizontalOffset) > threshold || Math.abs(verticalOffset) > threshold * 2;

    if (lookingAway) {
      logger.info(`Looking away detected: horizontal=${gazeData.horizontal_gaze}, vertical=${gazeData.vertical_gaze}`);
    }

    return { lookingAway, gazeData };
  } catch (error) {
    logger.error(`Eye tracking error: ${error.message}`);
    return { lookingAway: false, gazeData: null };
  }
}

/**
 * Check if eyes are closed
 * @param {Array} landmarks - Face landmarks
 * @returns {boolean}
 */
function areEyesClosed(landmarks) {
  const result = detectEyesLookingAway(null, landmarks);
  return result.gazeData?.eyes_closed || false;
}

module.exports = {
  detectEyesLookingAway,
  areEyesClosed,
  calculateEAR,
  GAZE_THRESHOLD,
};
